### PR TITLE
[SPARK-25402][SQL] Null handling in BooleanSimplification

### PR DESCRIPTION
**Pull request by gatorsmile**
## What changes were proposed in this pull request?
This PR is to fix the null handling in BooleanSimplification. In the rule BooleanSimplification, there are two cases that do not properly handle null values. The optimization is not right if either side is null. This PR is to fix them. 

## How was this patch tested?
Added test cases

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @cloud-fan @adrian-ionescu

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/BooleanSimplificationSuite.scala [hunk 1]
```diff
@@ -48,6 +48,14 @@ class BooleanSimplificationSuite extends PlanTest with PredicateHelper {
     testRelation.output, Seq(Row(1, 2, 3, "abc"))
   )
 
+  val testNotnullableRelation = LocalRelation('a.int.notNull, 'b.int.notNull, 'c.int.notNull,
```
Comment (gengliangwang) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/BooleanSimplificationSuite.scala]: Nit: `testNotnullableRelation` =>  `testNotNullableRelation`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/BooleanSimplificationSuite.scala [hunk 2]
```diff
@@ -48,6 +48,14 @@ class BooleanSimplificationSuite extends PlanTest with PredicateHelper {
...
+    'd.string.notNull, 'e.boolean.notNull, 'f.boolean.notNull, 'g.boolean.notNull,
+    'h.boolean.notNull)
+
+  val testNotnullableRelationWithData = LocalRelation.fromExternalRows(
```
Comment (gengliangwang) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/BooleanSimplificationSuite.scala]: `testNotnullableRelationWithData` => `testNotNullableRelationWithData`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 3]
```diff
@@ -263,10 +263,12 @@ object BooleanSimplification extends Rule[LogicalPlan] with PredicateHelper {
       case TrueLiteral Or _ => TrueLiteral
       case _ Or TrueLiteral => TrueLiteral
 
-      case a And b if Not(a).semanticEquals(b) => FalseLiteral
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: how about
```
case a And b if Not(a).semanticEquals(b) || a.semanticEquals(Not(b)) =>
  If(IsNull(a), null, FalseLiteral)
```

**Review by gengliangwang — APPROVED** [review]
Otherwise LGTM 👍

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala (hunk 3 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: +1

Comment (adrian-ionescu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: +1

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 4]
```diff
@@ -263,10 +263,12 @@ object BooleanSimplification extends Rule[LogicalPlan] with PredicateHelper {
       case TrueLiteral Or _ => TrueLiteral
```
Comment (adrian-ionescu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: While you're at it, isn't this also an issue?
```
Seq(false, true).toDF("b")
  .withColumn("b_or_null", 'b || lit(null))
  .withColumn("null_or_b", lit(null) || 'b)
  .withColumn("b_and_null", 'b && lit(null))
  .withColumn("null_and_b", lit(null) && 'b).show
+-----+---------+---------+----------+----------+
|    b|b_or_null|null_or_b|b_and_null|null_and_b|
+-----+---------+---------+----------+----------+
|false|     null|     null|     false|     false|
| true|     true|     true|      null|      null|
+-----+---------+---------+----------+----------+

```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala (hunk 3 above)
Comment (gatorsmile) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: How about 
```
      case a And b if Not(a).semanticEquals(b) || a.semanticEquals(Not(b)) =>
        if (!a.nullable && !b.nullable) {
          FalseLiteral
        } else {
          If(IsNull(a), Literal.create(null, a.dataType), FalseLiteral)
        }
```

Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: I think it should be 
```
      case a And b if Not(a).semanticEquals(b) || a.semanticEquals(Not(b)) =>
        if (!a.nullable && !b.nullable) {
          FalseLiteral
        } else if(a.nullable) {
          If(IsNull(a), Literal.create(null, a.dataType), FalseLiteral)
        } else {
          If(IsNull(b), Literal.create(null, a.dataType), FalseLiteral)
       }
```

**Comment by cloud-fan** [comment]
LGTM

**Review by gengliangwang — APPROVED** [review]
LGTM

**Review by adrian-ionescu — APPROVED** [review]
LGTM

**Comment by gengliangwang** [comment]
retest this please.

**Pull request closed without merging**